### ISSUE TITLE
Micro Frontend - A platform with a different configuration has been created. Please destroy it first

### ISSUE THREAD

**Issue by saikumar-madana** [angular-architects/module-federation-plugin#47]
I am trying to build a POC for Micro Frontend using Angular and Webpack Module Federation. In this, I have created a shell app and another mfe1 app and rendering that mfe1 on a specific route hit and it is working as expected and rendering the app. Now, I am trying to create another app called mfe2 and render it. In this mfe2, I am creating a web component using Angular elements and rendering that in the shell app. When I do that, I am facing the following issue

`Error: A platform with a different configuration has been created. Please destroy it first.`

![image](https://user-images.githubusercontent.com/17292167/118482090-e64cb780-b731-11eb-9a51-041246338bc9.png)

when the following code is executing

```
import('mfe2/web-component')
    .then(_ => console.debug(`element loaded!`))
    .catch(err => console.error(`error loading `, err));
```
I don't understand where the exact issue is. I am adding the required code below.

### MFE2:

app.module.ts
```
@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    AppRoutingModule
  ],
  providers: [],
  bootstrap: []
})
export class AppModule {
  constructor(private injector: Injector) {}

  ngDoBootstrap() {
    const ce = createCustomElement(AppComponent, {injector: this.injector});
    customElements.define('mfe2-elem', ce);
  }

}
```
webpack.config.js
```
module.exports = {
  output: {
    uniqueName: "mfe2",
    publicPath: "auto"
  },
  optimization: {
    runtimeChunk: false
  },   
  resolve: {
    alias: {
      ...sharedMappings.getAliases(),
    }
  },
  plugins: [
    new ModuleFederationPlugin({
      
        // For remotes (please adjust)
        name: "mfe2",
        filename: "remoteEntry.js",
        exposes: {
          './web-component': './src/bootstrap.ts',
        },        
        
        // For hosts (please adjust)
        // remotes: {
        //     "mfe1": "mfe1@http://localhost:3000/remoteEntry.js",

        // },

        shared: {
          "@angular/core": { singleton: true, strictVersion: true }, 
          "@angular/common": { singleton: true, strictVersion: true }, 
          "@angular/common/http": { singleton: true, strictVersion: true }, 
          "@angular/router": { singleton: true, strictVersion: true },

          ...sharedMappings.getDescriptors()
        }
        
    }),
    sharedMappings.getPlugin()
  ],
};
```

### SHELL:

rendering it in a component:
```
@Component({
  selector: 'app-mfe2element',
  templateUrl: './mfe2element.component.html',
  styleUrls: ['./mfe2element.component.scss']
})
export class Mfe2elementComponent implements AfterContentInit {

  @ViewChild('vc', { read: ElementRef, static: true })
  vc!: ElementRef;

  constructor() { }

  ngAfterContentInit(): void {
    import('mfe2/web-component')
    .then(_ => console.debug(`element loaded!`))
    .catch(err => console.error(`error loading `, err));

    // await import('mfe1/web-components');
    // const element = document.createElement('mfe1-element');
    // document.body.appendChild(element);

    const element = document.createElement('mfe2-elem');
    this.vc.nativeElement.appendChild(element);

  }

}
```
webpack.config.js
```
module.exports = {
  output: {
    uniqueName: "shell"
  },
  optimization: {
    // Only needed to bypass a temporary bug
    runtimeChunk: false
  },
  plugins: [
    new ModuleFederationPlugin({
        // For remotes (please adjust)
        // name: "shell",
        // filename: "remoteEntry.js",
        // exposes: {
        //     './Component': './/src/app/app.component.ts',
        // },        
        
        // For hosts (please adjust)
        remotes: {
            "mfe1": "mfe1@http://localhost:3000/remoteEntry.js",
            "mfe2": "mfe2@http://localhost:4000/remoteEntry.js",
        },

        shared: {
          "@angular/core": { singleton: true, strictVersion: false }, 
          "@angular/common": { singleton: true, strictVersion: false }, 
          "@angular/router": { singleton: true, strictVersion: false },

          ...sharedMappings.getDescriptors()
        }
        
    }),
    sharedMappings.getPlugin(),
  ],
};
```
Please help me solve this issue.

Thanks...

**Comment by kbleeck** [comment]
From https://www.angulararchitects.io/aktuelles/multi-framework-and-version-micro-frontends-with-module-federation-the-good-the-bad-the-ugly/:

> Per shared Angular version, we are only allowed to create one platform. To remember that there is already a shared platform for our version, we could put it into a global dictionary mapping the version number to the platform instance:

```declare const require: any;
const ngVersion = require('../package.json').dependencies['@angular/core']; // better just take the major version 

(window as any).plattform = (window as any).plattform || {};
let platform = (window as any).plattform[ngVersion];
if (!platform) {
  platform = platformBrowser();
  (window as any).plattform[ngVersion] = platform; 
}
platform.bootstrapModule(AppModule)
  .catch(err => console.error(err));
```

Adding this to the bootstrap.ts in both the shell and mfe will make the mfe reuse an existing platform instead of trying to create a new instance, which will lead to the error you're experiencing.

**Comment by saikumar-madana** [comment]
Thanks, @kbleeck that worked.

**Comment by manfredsteyer** [comment]
Thanks, @kbleeck for this pointer.

**Comment by rikersbeard** [comment]
Was this remedied for later versions?

**Comment by bh3605** [comment]
This was the solution I've came up with working in Angular 17. I needed to be able to export individual web components from my mfe. My setup is a monorepo of applications and libraries so they'll "always" use the same Angular version.

```
 
@NgModule({
    imports: [BrowserModule],
    providers: [
            {
              provide: ENVIRONMENT,
              useFactory: resolveEnvironment
            },
            getGlobalRoutes(),
            {
              provide: Router,
              useValue: getGlobalRouter()
            }
    ]
})
export class WebComponentsModule {
    constructor(private injector: EnvironmentInjector) {
       
    }
 
    ngDoBootstrap() {
        const recentClaims = createCustomElement(RecentClaimsComponent, {
            injector: this.injector
        });
        customElements.define('recent-claims', recentClaims);
    }
}

let platform = getPlatform();
platform?.bootstrapModule(WebComponentsModule).catch((err: any) => console.error(err));

```

Whether you are or are not sharing platforms, I don't see why getPlatform() would ever get you the wrong one. I did find the following also works. I don't know the consequences of setting the appType to either "microfrontend" or "shell". The "microfrontend" option establishes a window event listener when the router navigates. The "shell" option does something with NgZone.

```
bootstrap(WebComponentModule, {
    production: true,
    appType: "microfrontend",
    platformFactory:  () => getPlatform() as PlatformRef
});
```
When your mfe exposes multiple modules/web components/components then you'll probably going to see an error thrown about multiple platforms being created when you follow kbleek's example. The platform is already created. Go get it and use it. If you're exposing a single endpoint then I'd recommend kbleek's example else I'd recommend my first code example. However, if you're creating a modern Angular application (16+) then you're going to be using stand-alone components so kbleek's example can't be used anymore because you're calling `bootstrapApplication()` and not `platformBrowserDynamic().bootstrapModule()`.

If you were using different Angular versions then I'd highly recommend you pass in the Angular version into the bootstrap method. There's a VERSION property that's available from `@angular/core`. Then again I'm sure `getPlatform()` works just as fine so maybe there's a reason to not use the `bootstrap()` method anymore in these situations.
